### PR TITLE
Add option to refresh gcp token when config is cmd-path

### PR DESCRIPTION
**Pull request by jfrabaute**
The python client does not deal with "gcloud" gke auth in the kube config file as it's based on the "cmd-path" auth.

This is adding the code to take care of this auth.
This should solve kubernetes-client/python#754

### CODE REVIEW

**Comment by jfrabaute** [comment]
Looks like the test is failing on python style code problems on unrelated files (not thouched in that review).
I fixed a few, but I have one that I don't know about:

/tmp/tmp.0CpklmniPV/python/kubernetes/utils/quantity.py:24:80: E501 line too long (91 > 79 characters)

**Comment by micw523** [comment]
The CI is failing on the files you modified. Take a look at the diff output - or to save time, just use ```autopep8 --aggressive --aggressive --in-place kube_config.py```.

E501 is ignored in judging the success of the CI task.

**Comment by jfrabaute** [comment]
> The CI is failing on the files you modified. Take a look at the diff output - or to save time, just use `autopep8 --aggressive --aggressive --in-place kube_config.py`.
> 

Oops, yes. Fixed, build seems to pass now.

**Comment by jfrabaute** [comment]
Hi,

Any chance to have this reviewed?

File: config/kube_config.py [hunk 1]
```diff
@@ -156,7 +199,38 @@ def __init__(self, config_dict, active_context=None,
...
+                    '(cmd-path is an empty string in your kubeconfig file)')
+            if 'scopes' in config and config['scopes'] != "":
+                raise ConfigException(
+                    'scopes can only be used '
+                    'when kubectl is using a gcp service account key')
+            args = []
+            if 'cmd-args' in config:
+                args = config['cmd-args'].split()
+            else:
+                fields = config['cmd-path'].split()
```
Comment (yliaog) [inline config/kube_config.py]: this is assuming config['cmd-path'] exists, probably better to check to ensure that.

File: config/kube_config.py [hunk 2]
```diff
@@ -133,6 +135,47 @@ def as_data(self):
         return self._data
 
 
+class CommandTokenSource(object):
+    def __init__(self, cmd, args, tokenKey, expiryKey):
+
```
Comment (yliaog) [inline config/kube_config.py]: why the extra blank line here?

File: config/kube_config_test.py [hunk 3]
```diff
@@ -727,7 +731,8 @@ class TestKubeConfigLoader(BaseTestCase):
                     "auth-provider": {
                         "config": {
                             "access-token": TEST_AZURE_TOKEN,
-                            "apiserver-id": "00000002-0000-0000-c000-000000000000",
+                            "apiserver-id": "00000002-0000-0000-c000-"
```
Comment (yliaog) [inline config/kube_config_test.py]: what is the reason for this change?

File: config/kube_config_test.py [hunk 4]
```diff
@@ -1439,7 +1444,8 @@ def test_list_kube_config_contexts(self):
             {'context': {'cluster': 'ssl', 'user': 'ssl'}, 'name': 'ssl'},
             {'context': {'cluster': 'default', 'user': 'simple_token'},
              'name': 'simple_token'},
-            {'context': {'cluster': 'default', 'user': 'expired_oidc'}, 'name': 'expired_oidc'}]
+            {'context': {'cluster': 'default', 'user': 'expired_oidc'},
+             'name': 'expired_oidc'}]
```
Comment (yliaog) [inline config/kube_config_test.py]: please add a test case

File: config/kube_config.py (hunk 1 above)
Comment (jfrabaute) [reply config/kube_config.py]: It should be set, otherwise it would have failed before (line 204 at least).
And this function is also called when 'cmd-path' is set so it should be set.

File: config/kube_config.py (hunk 2 above)
Comment (jfrabaute) [reply config/kube_config.py]: Fixed (when PR will be updated)

File: config/kube_config_test.py (hunk 3 above)
Comment (jfrabaute) [reply config/kube_config_test.py]: The code is exactly the same, but I had to split the string into 2 lines because some styling tests were failing with a line too long.
It's probably not related to my change, and it might be a styling check which has been added after this file was in the repo (not sure).
Maybe it should be a preliminary separate commit, not sure.

File: config/kube_config_test.py (hunk 4 above)
Comment (jfrabaute) [reply config/kube_config_test.py]: Working on it.

Comment (jfrabaute) [reply config/kube_config_test.py]: Added tests and updated the PR.

File: config/kube_config_test.py [hunk 5]
```diff
@@ -1394,6 +1501,7 @@ class TestKubeConfigMerger(BaseTestCase):
             {
                 "name": "expired_oidc",
                 "user": {
+                    "name": "gcp",
```
Comment (yliaog) [inline config/kube_config_test.py]: why adding it? it does not relate to your code change, right?

File: config/kube_config_test.py [hunk 6]
```diff
@@ -1279,6 +1346,46 @@ def test_user_exec_auth(self, mock):
...
+        token = "dummy"
+        return_value = A(token, parse_rfc3339(datetime.datetime.now()))
+        CommandTokenSource.token = mock.Mock(return_value=return_value)
+        expected = FakeConfig(api_key={
+                              "authorization": BEARER_TOKEN_FORMAT % token})
+        actual = FakeConfig()
+        KubeConfigLoader(
+            config_dict=self.TEST_KUBE_CONFIG,
+            active_context="contexttestcmdpath").load_and_set(actual)
+        expected.get_api_key_with_prefix = actual.get_api_key_with_prefix
```
Comment (yliaog) [inline config/kube_config_test.py]: why? could you assert expected.api_key instead?

File: config/kube_config_test.py [hunk 7]
```diff
@@ -1279,6 +1346,46 @@ def test_user_exec_auth(self, mock):
...
+
+    def test_user_cmd_path_empty(self):
+        A = namedtuple('A', ['token', 'expiry'])
+        token = "dummy"
+        return_value = A(token, parse_rfc3339(datetime.datetime.now()))
+        CommandTokenSource.token = mock.Mock(return_value=return_value)
+        expected = FakeConfig(api_key={
+                              "authorization": BEARER_TOKEN_FORMAT % token})
+        actual = FakeConfig()
+        with self.assertRaises(ConfigException):
```
Comment (yliaog) [inline config/kube_config_test.py]: better to also check the raised exception is the correct one

File: config/kube_config_test.py [hunk 8]
```diff
@@ -1279,6 +1346,46 @@ def test_user_exec_auth(self, mock):
...
+
+    def test_user_cmd_path_with_scope(self):
+        A = namedtuple('A', ['token', 'expiry'])
+        token = "dummy"
+        return_value = A(token, parse_rfc3339(datetime.datetime.now()))
+        CommandTokenSource.token = mock.Mock(return_value=return_value)
+        expected = FakeConfig(api_key={
+                              "authorization": BEARER_TOKEN_FORMAT % token})
+        actual = FakeConfig()
+        with self.assertRaises(ConfigException):
```
Comment (yliaog) [inline config/kube_config_test.py]: better to also check the raised exception is the correct one

File: config/kube_config_test.py (hunk 5 above)
Comment (jfrabaute) [reply config/kube_config_test.py]: Ooops, my bad, fixed.

File: config/kube_config_test.py (hunk 7 above)
Comment (jfrabaute) [reply config/kube_config_test.py]: Done.

File: config/kube_config_test.py (hunk 8 above)
Comment (jfrabaute) [reply config/kube_config_test.py]: Done.

File: config/kube_config_test.py (hunk 6 above)
Comment (jfrabaute) [reply config/kube_config_test.py]: This is brekaing the "self.assertEqual" on the next line, because it looks like this attribute is transient and then it should not be part of the assertEqual, but right now this is how it's done, so I set this field as well, so the assert succeeds, but as this field is not important for the test, I just set it.
Is there a nicer way (I agree that it's ugly).

Comment (yliaog) [reply config/kube_config_test.py]: what about clear get_api_key_with_prefix in actual?

Comment (jfrabaute) [reply config/kube_config_test.py]: Done, Updated PR

**Comment by jfrabaute** [comment]
Thanks @yliaog for the reviews.
I'll work on the other PR and add tests now: https://github.com/kubernetes-client/python-base/pull/169

**Comment by palnabarun** [comment]
Note to self: We should add documentation for the functionality implemented here.

/cc

cc: @roycaihw
